neurolib/models/multimodel/builder/adex.py

```diff
@@ -0,0 +1,944 @@
+        if any(k in params_dict for k in ("c_global", "J_exc_max", "tau_syn_exc")):
+            # get original c_global
+            c_global = (
+                params_dict["c_global"]
+                if "c_global" in params_dict
+                else self.params["c_global"] * (self.params["J_exc_max"] / self.params["tau_syn_exc"])
+            )
+            tau_syn_exc = params_dict["tau_syn_exc"] if "tau_syn_exc" in params_dict else self.params["tau_syn_exc"]
+            J_exc_max = params_dict["J_exc_max"] if "J_exc_max" in params_dict else self.params["J_exc_max"]
+            params_dict["c_global"] = c_global * tau_syn_exc / J_exc_max
+
+        # update all parameters finally
+        super().update_params(params_dict)
+
+    def describe(self):
+        return {
+            **super().describe(),
+            "lin_nonlin_cascade_filename": self.lin_nonlin_fname,
+        }
+
+    def _callbacks(self):
+        """
+        Construct list of python callbacks for AdEx model.
+        """
+        callbacks_list = [
+            (self.callback_functions["firing_rate_lookup"], self.firing_rate_lookup, 2),
+            (self.callback_functions["voltage_lookup"], self.voltage_lookup, 2),
+            (self.callback_functions["tau_lookup"], self.tau_lookup, 2),
+        ]
+        self._validate_callbacks(callbacks_list)
+        return callbacks_list
+
+    def _numba_callbacks(self):
+        """
+        Define numba callbacks - has to be different than jitcdde callbacks
+        because of the internals.
+        """
+
+        def _table_numba_gen(sigma_range, mu_range, d_sigma, d_mu, cascade):
+            """
+            Function generator for numba callbacks. This works similarly as
+            `functools.partial` (i.e. sets some of the arguments of the inner
+            function), but afterwards can be jitted with `numba.njit()`, while
+            partial functions cannot.
+            """
+
+            def inner(current_mu, current_sigma):
+                return _table_lookup(current_mu, current_sigma, sigma_range, mu_range, d_sigma, d_mu, cascade,)
+
+            return inner
+
+        return [
+            (
+                "firing_rate_lookup",
+                numba.njit(
+                    _table_numba_gen(
+                        self.sigma_range, self.mu_range, self.d_sigma, self.d_mu, self.firing_rate_cascade,
+                    )
+                ),
+            ),
+            (
+                "voltage_lookup",
+                numba.njit(
+                    _table_numba_gen(self.sigma_range, self.mu_range, self.d_sigma, self.d_mu, self.voltage_cascade)
+                ),
+            ),
+            (
+                "tau_lookup",
+                numba.njit(
+                    _table_numba_gen(self.sigma_range, self.mu_range, self.d_sigma, self.d_mu, self.tau_cascade)
+                ),
+            ),
+        ]
+
+    def firing_rate_lookup(self, y, current_mu, current_sigma):
```

Comment:
this is the actual definition of first callback function - table lookup for firing rate